### PR TITLE
Feat/layer agg

### PR DESCRIPTION
**Pull request by y-prudent**
**Overview**

This PR introduces multi-layer scoring for OOD detectors. To achieve this, a new abstract class, `Aggregator`, has been implemented. This class is trained via its `fit` method on in-distribution scores collected from various layers and is later used to aggregate new scores using its `aggregate` method. The aggregator is instantiated during the initialization of an OOD detector and is invoked within both the `fit` and `score` methods. In addition, all distance-based OOD detection methods have been updated to incorporate the aggregator feature.

**Minor changes**

- Added default post-processing functions `_default_postproc_fn` for both PyTorch and TensorFlow that include average pooling, robust to different feature shapes (respectively in `oodeel/extractor/torch_feature_extractor.py` and `oodeel/extractor/keras_feature_extractor.py`).
- The `postproc_fns` parameter is now accepted by `OODBaseDetector.fit()` (instead of `__init__`) to mirror how `feature_layers_ids` is handled (`oodeel/methods/base.py`).
- Updated the `pylint` repository URL in `.pre-commit-config.yaml` to reflect the new official repository (`https://github.com/pylint-dev/pylint`) and changed the version from `v3.0.0a3` to `v3.3.6`. 

**New Features**

- Added a `BaseAggregator` class in `oodeel/aggregator/base.py` to define the interface for aggregating OOD detection scores. This serves as the foundation for other aggregator implementations.
- Introduced the `FisherAggregator` class in `oodeel/aggregator/fisher.py`, which uses Fisher's method with Brown's correction to combine OOD scores from multiple feature layers (from Combine and Conquer approach).
- Added the `MeanNormalizedAggregator` class in `oodeel/aggregator/mean.py`, which normalizes per-layer scores by their mean before aggregating them.
- Added the `StdNormalizedAggregator` class in `oodeel/aggregator/std.py`, which normalizes per-layer scores by their standard deviation before aggregation.
- Updated `DKNN` to support score aggregation across multiple layers (`oodeel/methods/dknn.py`).
- Updated `Mahalanobis` to support score aggregation across multiple layers (`oodeel/methods/mahalanobis.py`).
- Updated `RMDS` to support score aggregation across multiple layers (`oodeel/methods/rmds.py`).
- Updated `Gram` to support score aggregation across multiple layers (`oodeel/methods/gram.py`).
- Updated `VIM` to support score aggregation across multiple layers (`oodeel/methods/vim.py`).
- Updated `SHE` to support score aggregation across multiple layers (`oodeel/methods/she.py`).

**TODO**

- [x]  Add tests for aggregator implementations (Fisher, MeanNormalized, StdNormalized). => in DKNN tests
- [x] Refactor Maha, Gram, RMDS, VIM to get per layer fit and score functions
- [x] Fix Gram corner case of too small validation set with no samples for some classes
- [x] Remove input perturbation in RMDS ?? 🤔 => Kept input perturbation but wrt softmax (as in ODIN)
- [x] SHE input perturbation wrt softmax score instead of perturbation wrt SHE score?? Same for RMDS?? 🤔 
- [x]  Update all feature based methods notebooks.

### CODE REVIEW

File: oodeel/methods/dknn.py [hunk 1]
```diff
@@ -56,54 +69,174 @@ def __init__(self, nearest: int = 50, use_gpu: bool = False):
...
+                    aggregator is used; otherwise, None.
+        """
+        norm_features = self._prepare_layer_features(layer_features)
+        index = self._create_index(norm_features.shape[1])
+        index.add(norm_features)
+
+        scores = None
+        if self.aggregator is not None:
+            # Use only a subset of samples for computing initial scores, ensuring at
+            # least 2 neighbors (distance to self is 0)
```
Comment (cofri) [inline oodeel/methods/dknn.py:128]: Why is it required for dknn? Maybe we could add a more descriptive comment?

File: oodeel/methods/dknn.py [hunk 2]
```diff
@@ -56,54 +69,174 @@ def __init__(self, nearest: int = 50, use_gpu: bool = False):
...
-        norm_input_projected = self._l2_normalization(input_projected)
-        scores, _ = self.index.search(norm_input_projected, self.nearest)
-        return scores[:, -1]
+        for i, index in enumerate(self.indexes):
+            scores.append(self._score_layer(index, input_projected[i]))
+
+        if self.aggregator is not None:
+            aggregated_scores = self.aggregator.aggregate(scores)
+        elif len(scores) == 1:
+            aggregated_scores = scores[0]
```
Comment (cofri) [inline oodeel/methods/dknn.py:224]: Same question as above: what if the user defines an aggregator but with a single feature_layer?

**Review by cofri — COMMENTED** [review]
Big and nice work!
I didn't have time to review all files. But here is a first pass.

**Review by paulnovello — CHANGES_REQUESTED** [review]
Very good job! I really like the new structure of OOD detector classes. Super idea to use default postropc_fn since it is rare not to pool internal features.

You can find some minor comments in my review but I have a more general one: would it be possible to rearrange the code a little bit so that `score_tensor` and `fit_to_dataset` become generic OODBaseDetector methods that only rely on `score_layer` and `fit_layer`?

**Review by paulnovello — COMMENTED** [review]
Not sure that you updated the Readme and index.md

File: oodeel/aggregator/fisher.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        Args:
+            scores (np.ndarray): The stacked scores (ID or test) with shape
+                (num_samples, num_layers).
+
+        Returns:
+            np.ndarray: A 1D array of Fisher combined test statistics, one for each
+                sample.
+        """
+        p_values = self._compute_p_values(scores)
+        return fisher_tau_method(p_values)
```
[the remark this answers is not in the document]
Comment (y-prudent) [reply oodeel/aggregator/fisher.py:131]: You are right, everything is now included in the class as private methods. For better clarity I reordered the methods with first public api methods, then private methods

File: oodeel/methods/dknn.py (hunk 1 above)
Comment (y-prudent) [reply oodeel/methods/dknn.py:128]: When fitting the aggregator, we use a subset of the training (fit) set to compute OOD scores. However, computing KNN scores with k=1 on the same samples used for fitting leads to zero-distance scores (since each sample is its own nearest neighbor), which is not meaningful for the aggregator.

To avoid this issue, we ensure that k >= 2 when computing KNN scores used to fit the aggregator, so that each sample’s score reflects a true distance to other points in the fit set.

File: oodeel/methods/dknn.py [hunk 4]
```diff
@@ -56,54 +69,174 @@ def __init__(self, nearest: int = 50, use_gpu: bool = False):
...
         """
-        fit_projected, _ = self.feature_extractor.predict(fit_dataset)
-        fit_projected = self.op.convert_to_numpy(fit_projected[0])
-        fit_projected = fit_projected.reshape(fit_projected.shape[0], -1)
-        norm_fit_projected = self._l2_normalization(fit_projected)
+        features_np = self.op.convert_to_numpy(features)
+        features_np = features_np.reshape(features_np.shape[0], -1)
+        return self._l2_normalization(features_np)
 
+    def _create_index(self, dim: int) -> faiss.IndexFlatL2:
```
[the remark this answers is not in the document]
Comment (y-prudent) [reply oodeel/methods/dknn.py:86]: Good point, I reorganized an tried to get something consistent among the different feature based methods

**Comment by y-prudent** [comment]
Ready to be reviewed! Here is a summary of the updates:

- **Reorder methods** in feature-based approaches for better clarity
- **Refactorization:** _fit_to_dataset and _score_tensor are now defined in a new class `FeatureBasedDetector`. Only `_fit_layer` and `score_layer` are required in the detector classes.
- **Input perturbation**: Now input perturbation (from ODIN) is available for every baseline through the initialization arguments `eps` and `temperature` (run only if eps > 0)

File: tests/tests_torch/torch_methods_utils.py [hunk 5]
```diff
@@ -94,7 +94,10 @@ def eval_detector_on_blobs(
 
     # fit ood detector
     if detector.requires_to_fit_dataset or detector.use_react:
-        detector.fit(model, feature_layers_id=[-2], fit_dataset=ds_fit)
+        if hasattr(detector, "aggregator") and detector.aggregator is not None:
+            detector.fit(model, feature_layers_id=[-2, -1], fit_dataset=ds_fit)
+        else:
+            detector.fit(model, feature_layers_id=[-2], fit_dataset=ds_fit)
```
[the remark this answers is not in the document]
Comment (y-prudent) [reply tests/tests_torch/torch_methods_utils.py:100]: Not relevant

File: oodeel/methods/dknn.py (hunk 2 above)
Comment (y-prudent) [reply oodeel/methods/dknn.py]: Fixed in `FeatureBasedDetector._score_tensor` function

File: oodeel/methods/she.py [hunk 6]
```diff
@@ -39,142 +41,116 @@ class SHE(OODBaseDetector):
...
-            inputs (TensorType): input samples
+            layer_features: Tensor of shape `(N, D)` containing the flattened
+                activations of in-distribution samples for one layer.
+            labels_np: Ground-truth class labels as a `np.ndarray`.
+            preds_np: Model predictions for the in-distribution samples.  Means
+                are only computed from samples that are *both* predicted and
+                labelled as the target class (following the original paper).
+            subset_size: Number of samples on which to compute provisional OOD
+                scores for the aggregator fit.  Ignored if *self.aggregator* is
+                *None*.
```
Comment (paulnovello) [inline oodeel/methods/she.py:100]: docstring not consistent

File: oodeel/methods/rmds.py [hunk 7]
```diff
@@ -20,103 +20,169 @@
...
+        super().__init__(
+            eps=eps, temperature=temperature, aggregator=aggregator, **kwargs
+        )
+        # Will be filled by `_fit_layer`.
+        self._layer_background_stats: List[Tuple[TensorType, np.ndarray]] = []
+
+    # === Per-layer logic ===
+    def _fit_layer(
+        self,
+        layer_id: int,
```
Comment (paulnovello) [inline oodeel/methods/rmds.py:69]: Not used, can remove (check for all detectors)

File: oodeel/methods/vim.py [hunk 8]
```diff
@@ -20,184 +20,242 @@
...
+        self.alphas.append(alpha)
+
+        return scores
+
+    def _score_layer(
+        self,
+        layer_id: int,
+        layer_features: TensorType,
+        info: dict,
+        **kwargs,
```
Comment (paulnovello) [inline oodeel/methods/vim.py:162]: remove kwargs?

File: oodeel/methods/rmds.py (hunk 7 above)
Comment (y-prudent) [reply oodeel/methods/rmds.py:69]: True but in some other detectors this parameter is used, and when `fit_layer` is called in `FeatureBasedDetectorpass`'s `_fit_to_dataset` method, the `layer_id` parameter is expected.

**Review by paulnovello — APPROVED** [review]
LGTM, great work !!
